File: shared/events/event_bus.py
from typing import Callable, Dict, List, Any, Optional
from shared.events.event_types import Event


class EventBus:
    """Lightweight in-memory event bus supporting pub/sub, once, clear, and history."""

    def __init__(self, debug_mode: bool = False, history_limit: int = 1000):
        self._listeners: Dict[str, List[Callable[[Event], None]]] = {}
        self._once_listeners: Dict[str, List[Callable[[Event], None]]] = {}
        self.debug_mode = debug_mode
        self.history_limit = history_limit
        self.event_history: List[Dict[str, Any]] = []

    def subscribe(self, event_name: str, handler: Callable[[Event], None]) -> None:
        if event_name not in self._listeners:
            self._listeners[event_name] = []
        if handler not in self._listeners[event_name]:
            self._listeners[event_name].append(handler)

    def unsubscribe(self, event_name: str, handler: Callable[[Event], None]) -> bool:
        removed = False
        if event_name in self._listeners and handler in self._listeners[event_name]:
            self._listeners[event_name].remove(handler)
            removed = True
        if event_name in self._once_listeners and handler in self._once_listeners[event_name]:
            self._once_listeners[event_name].remove(handler)
            removed = True
        return removed

    def once(self, event_name: str, handler: Callable[[Event], None]) -> None:
        if event_name not in self._once_listeners:
            self._once_listeners[event_name] = []
        if handler not in self._once_listeners[event_name]:
            self._once_listeners[event_name].append(handler)

    def publish(self, event_name: str, data: Any = None) -> Event:
        event = data if isinstance(data, Event) else Event(name=event_name, data=data)

        if self.debug_mode:
            self.event_history.append(event.to_dict())
            if len(self.event_history) > self.history_limit:
                self.event_history.pop(0)

        # Notify persistent listeners
        if event_name in self._listeners:
            for handler in list(self._listeners[event_name]):
                try:
                    handler(event)
                except Exception as e:
                    print(f"Error executing listener for {event_name}: {e}")

        # Notify once listeners and clear them
        if event_name in self._once_listeners:
            listeners = list(self._once_listeners[event_name])
            self._once_listeners[event_name].clear()
            for handler in listeners:
                try:
                    handler(event)
                except Exception as e:
                    print(f"Error executing once listener for {event_name}: {e}")

        return event

    def clear(self, event_name: Optional[str] = None) -> None:
        if event_name:
            if event_name in self._listeners:
                self._listeners[event_name].clear()
            if event_name in self._once_listeners:
                self._once_listeners[event_name].clear()
        else:
            self._listeners.clear()
            self._once_listeners.clear()
            self.event_history.clear()

    def get_listener_count(self, event_name: str) -> int:
        count = len(self._listeners.get(event_name, []))
        count += len(self._once_listeners.get(event_name, []))
        return count

File: static/assets/js/utils/event_bus.js
/**
 * EventBus
 * Production-grade, lightweight in-memory event bus for PostureSense v2 browser engine pipeline.
 */

export class EventBus {
    constructor() {
        this.listeners = new Map();
        this.wildcardListeners = new Set();
    }

    /**
     * Subscribe to an event topic or all events ('*')
     * @param {string} eventName
     * @param {function} handler
     * @returns {function} Unsubscribe function
     */
    subscribe(eventName, handler) {
        if (typeof handler !== 'function') {
            console.error('[EventBus] Handler must be a function', handler);
            return () => {};
        }

        if (eventName === '*') {
            this.wildcardListeners.add(handler);
            return () => this.unsubscribe('*', handler);
        }

        if (!this.listeners.has(eventName)) {
            this.listeners.set(eventName, new Set());
        }
        this.listeners.get(eventName).add(handler);

        return () => this.unsubscribe(eventName, handler);
    }

    /**
     * Unsubscribe a handler from an event topic
     * @param {string} eventName
     * @param {function} handler
     */
    unsubscribe(eventName, handler) {
        if (eventName === '*') {
            this.wildcardListeners.delete(handler);
            return;
        }

        if (this.listeners.has(eventName)) {
            this.listeners.get(eventName).delete(handler);
            if (this.listeners.get(eventName).size === 0) {
                this.listeners.delete(eventName);
            }
        }
    }

    /**
     * Publish an event to subscribers
     * @param {string} eventName
     * @param {*} data
     */
    publish(eventName, data) {
        const eventObject = {
            name: eventName,
            data: data,
            timestamp: Date.now()
        };

        // Notify specific subscribers
        if (this.listeners.has(eventName)) {
            for (const handler of Array.from(this.listeners.get(eventName))) {
                try {
                    handler(eventObject);
                } catch (err) {
                    console.error(`[EventBus] Error executing subscriber for '${eventName}':`, err);
                }
            }
        }

        // Notify wildcard subscribers
        for (const handler of Array.from(this.wildcardListeners)) {
            try {
                handler(eventObject);
            } catch (err) {
                console.error(`[EventBus] Error executing wildcard subscriber for '${eventName}':`, err);
            }
        }
    }

    /**
     * Clear all subscribers
     */
    clear() {
        this.listeners.clear();
        this.wildcardListeners.clear();
    }
}
